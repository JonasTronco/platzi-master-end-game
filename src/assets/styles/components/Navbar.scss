@import '../main.scss';

.navbar {
  @include colCenCen;
  @include card;
  width: 175px;
  margin-right: 39px;
  padding: $space_x3;

  &__user {
    @include colCenCen;
    margin: $space_x3 0;

    &__photo {
      width: 120px;
      height: 120px;
      padding: 0;
      margin-bottom: $space_x2;
      border-radius: 50%;
      overflow: hidden;
      box-shadow: $shadow;

      &__img {
        width: 100%;
        height: 100%;
      }
    }

    &__name {
      text-align: center;
      font-size: $topic_size;
    }

    &__type {
      text-transform: capitalize;
    }
  }

  &__options {
    align-self: flex-start;
    padding: 0;
    list-style-type: none;

    &__item {
      @include flexStaCen;
      margin-bottom: $space_x3;
      cursor: pointer;

      &__icon {
        width: 20px;
        height: 20px;
        margin-right: 5px;
        fill: $secondary_color;
      }
    }
  }
}
