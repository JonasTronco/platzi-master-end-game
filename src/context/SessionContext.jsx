import { useState, useEffect, createContext, useMemo, useContext } from 'react';
import Router from 'next/router';

const SessionContext = createContext();
export function SessionProvider(props) {
  const [session, setSession] = useState(null);

  useEffect(() => {
    try {
      setSession(JSON.parse(localStorage.getItem('session')));
    } catch (error) {
      Router.push('/');
    }
  }, []);

  const value = useMemo(
    () => ({
      session,
    }),
    [session],
  );

  return <SessionContext.Provider value={value} {...props} />;
}

export function useSession() {
  const context = useContext(SessionContext);
  if (!context) {
    throw new Error('No Session Context Provider');
  }
  return context;
}
