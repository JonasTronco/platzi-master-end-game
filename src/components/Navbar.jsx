import Router from 'next/router';
import Link from 'next/link';
import PropTypes from 'prop-types';
import { useSession } from '../context/SessionContext';
import '../assets/styles/components/Navbar.scss';
import UserIcon from '../assets/icons/user.svg';
import UsersIcon from '../assets/icons/users.svg';
import UserAddIcon from '../assets/icons/user-add.svg';
import UsersAddIcon from '../assets/icons/users-add.svg';
import LogoutIcon from '../assets/icons/logout.svg';
import TestIcon from '../assets/icons/exam.svg';

function Navbar() {
  const { session } = useSession();

  function logout() {
    document.cookie = 'session=; expires=Thu, 01 Jan 1970 00:00:01 GMT;';
    Router.push('/');
  }

  function TypeOptions({ type }) {
    switch (type) {
      case 'Administrator':
        return (
          <>
            <Link href="/users">
              <li className="navbar__options__item">
                <UsersIcon className="navbar__options__item__icon" />
                <strong className="navbar__options__item__text">Users</strong>
              </li>
            </Link>
            <Link href="/add-user">
              <li className="navbar__options__item">
                <UserAddIcon className="navbar__options__item__icon" />
                <strong className="navbar__options__item__text">
                  Add User
                </strong>
              </li>
            </Link>
            <Link href="/add-users">
              <li className="navbar__options__item">
                <UsersAddIcon className="navbar__options__item__icon" />
                <strong className="navbar__options__item__text">
                  Add Users
                </strong>
              </li>
            </Link>
          </>
        );
      case 'Patient':
        return (
          <Link href="/tests">
            <li className="navbar__options__item">
              <TestIcon className="navbar__options__item__icon" />
              <strong className="navbar__options__item__text">Tests</strong>
            </li>
          </Link>
        );
      default:
        return (
          <Link href="/patients">
            <li className="navbar__options__item">
              <UsersIcon className="navbar__options__item__icon" />
              <strong className="navbar__options__item__text">Patients</strong>
            </li>
          </Link>
        );
    }
  }
  TypeOptions.propTypes = {
    type: PropTypes.string.isRequired,
  };

  if (session && session.user) {
    return (
      <section className="navbar">
        <div className="navbar__user">
          <figure className="navbar__user__photo">
            <img
              className="navbar__user__photo__img"
              src={
                session.user.imageURL
                  ? session.user.imageURL
                  : 'https://i.imgur.com/oMJFiLX.jpg'
              }
              alt="user"
            />
          </figure>
          <strong className="navbar__user__name">{`${session.user.firstName} ${session.user.lastName}`}</strong>
          <span className="navbar__user__type">{session.user.typeOfUser}</span>
        </div>
        <ul className="navbar__options">
          <li
            className="navbar__options__item"
            onClick={() => {
              Router.push(`/profile/${session.user.id}`);
            }}
          >
            <UserIcon className="navbar__options__item__icon" />
            <strong className="navbar__options__item__text">Profile</strong>
          </li>
          <TypeOptions type={session.user.typeOfUser} />
          <li className="navbar__options__item" onClick={logout}>
            <LogoutIcon className="navbar__options__item__icon" />
            <strong className="navbar__options__item__text">Logout</strong>
          </li>
        </ul>
      </section>
    );
  }

  return <div className="loader" />;
}

export default Navbar;
