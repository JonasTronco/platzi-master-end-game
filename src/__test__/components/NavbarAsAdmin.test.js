import Navbar from '../../components/Navbar';
import { create } from 'react-test-renderer';
import Router from 'next/router';
import { render, fireEvent, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import { SessionProvider } from '../../context/SessionContext';

describe('Navbar as Doctor Component', () => {
  const mockedRouter = { push: () => {} };
  Router.router = mockedRouter;
  const obj = {
    session: {
      token:
        'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiI1ZWM1Y2UxNmZjMTNhZTE1MDYwMDAwNmEiLCJ1c2VybmFtZSI6InRpZmYua2VuZGVyLjU2ODQiLCJzY29wZXMiOltdLCJpYXQiOjE1OTA3Mzg4MzAsImV4cCI6MTU5MTE3MDgzMH0.jIQdIXxhovhiLVjaQdEtEk1HbQreLmZuRWivKtnJocc',
      user: {
        id: '5ec5ce16fc13ae150600006a',
        username: 'tiff.kender.5684',
        typeOfUser: 'Administrator',
        isActive: true,
        imageURL: 'http://dummyimage.com/200x200.jpg/ff4444/ffffff',
        firstName: 'Tiff',
        lastName: 'Kender',
        defaultPath: '/users'
      }
    }
  };

  test('Should render Navbar component', () => {
    const component = create(
      <SessionProvider value={obj}>
        <Navbar />
      </SessionProvider>
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
});
