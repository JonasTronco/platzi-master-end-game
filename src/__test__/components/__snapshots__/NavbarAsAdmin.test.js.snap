// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar as Doctor Component Should render Navbar component 1`] = `
<section
  className="navbar"
>
  <div
    className="navbar__user"
  >
    <figure
      className="navbar__user__photo"
    >
      <img
        alt="user"
        className="navbar__user__photo__img"
        src="http://dummyimage.com/200x200.jpg/ff4444/ffffff"
      />
    </figure>
    <strong
      className="navbar__user__name"
    >
      Tiff Kender
    </strong>
    <span
      className="navbar__user__type"
    >
      Administrator
    </span>
  </div>
  <ul
    className="navbar__options"
  >
    <li
      className="navbar__options__item"
      onClick={[Function]}
    >
      <div
        className="navbar__options__item__icon"
      />
      <strong
        className="navbar__options__item__text"
      >
        Profile
      </strong>
    </li>
    <li
      className="navbar__options__item"
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      <div
        className="navbar__options__item__icon"
      />
      <strong
        className="navbar__options__item__text"
      >
        Users
      </strong>
    </li>
    <li
      className="navbar__options__item"
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      <div
        className="navbar__options__item__icon"
      />
      <strong
        className="navbar__options__item__text"
      >
        Add User
      </strong>
    </li>
    <li
      className="navbar__options__item"
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      <div
        className="navbar__options__item__icon"
      />
      <strong
        className="navbar__options__item__text"
      >
        Add Users
      </strong>
    </li>
    <li
      className="navbar__options__item"
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      <div
        className="navbar__options__item__icon"
      />
      <strong
        className="navbar__options__item__text"
      >
        Logout
      </strong>
    </li>
  </ul>
</section>
`;
